base/applications/rapps/settings.cpp

```diff
@@ -0,0 +1,247 @@
+{
+    BOOL bAllSuccess = TRUE;
+    ATL::CAtlList<SettingsField *> infoFields;
+
+    AddInfoFields(infoFields, settings);
+
+    POSITION InfoListPosition = infoFields.GetHeadPosition();
+    while (InfoListPosition)
+    {
+        SettingsField *Info = infoFields.GetNext(InfoListPosition);
+        if (!Info->Load(key))
+        {
+            bAllSuccess = FALSE;
+            // TODO: error log
+        }
+        delete Info;
+    }
+    return bAllSuccess;
+}
+
+VOID FillDefaultSettings(PSETTINGS_INFO pSettingsInfo)
+{
+    ATL::CStringW szDownloadDir;
+    ZeroMemory(pSettingsInfo, sizeof(SETTINGS_INFO));
+
+    pSettingsInfo->bSaveWndPos = TRUE;
+    pSettingsInfo->bUpdateAtStart = FALSE;
+    pSettingsInfo->bLogEnabled = TRUE;
+    pSettingsInfo->bUseSource = FALSE;
+
+    if (FAILED(SHGetFolderPathW(NULL, CSIDL_PERSONAL, NULL, SHGFP_TYPE_CURRENT, szDownloadDir.GetBuffer(MAX_PATH))))
+    {
+        szDownloadDir.ReleaseBuffer();
+        if (!szDownloadDir.GetEnvironmentVariableW(L"SystemDrive"))
+        {
+            szDownloadDir = L"C:";
+        }
+    }
+    else
+    {
+        szDownloadDir.ReleaseBuffer();
+    }
+
+    PathAppendW(szDownloadDir.GetBuffer(MAX_PATH), L"\\RAPPS Downloads");
+    szDownloadDir.ReleaseBuffer();
+
+    ATL::CStringW::CopyChars(pSettingsInfo->szDownloadDir,
+        _countof(pSettingsInfo->szDownloadDir),
+        szDownloadDir.GetString(),
+        szDownloadDir.GetLength() + 1);
+
+    pSettingsInfo->bDelInstaller = FALSE;
+    pSettingsInfo->Maximized = FALSE;
+    pSettingsInfo->Left = CW_USEDEFAULT;
+    pSettingsInfo->Top = CW_USEDEFAULT;
+    pSettingsInfo->Width = 680;
+    pSettingsInfo->Height = 450;
+}
+
+BOOL LoadSettings(PSETTINGS_INFO pSettingsInfo)
+{
+    ATL::CRegKey RegKey;
+    if (RegKey.Open(HKEY_CURRENT_USER, L"Software\\ReactOS\\rapps", KEY_READ) != ERROR_SUCCESS)
+    {
+        return FALSE;
+    }
+
+    BOOL bRet = LoadAllSettings(RegKey, *pSettingsInfo);
+    RegKey.Close();
+    return bRet;
+}
+
+BOOL SaveSettings(HWND hwnd, PSETTINGS_INFO pSettingsInfo)
+{
+    WINDOWPLACEMENT wp;
+    ATL::CRegKey RegKey;
+
+    if (pSettingsInfo->bSaveWndPos)
+    {
+        wp.length = sizeof(wp);
+        GetWindowPlacement(hwnd, &wp);
+
+        pSettingsInfo->Left = wp.rcNormalPosition.left;
+        pSettingsInfo->Top = wp.rcNormalPosition.top;
+        pSettingsInfo->Width = wp.rcNormalPosition.right - wp.rcNormalPosition.left;
+        pSettingsInfo->Height = wp.rcNormalPosition.bottom - wp.rcNormalPosition.top;
+        pSettingsInfo->Maximized = (wp.showCmd == SW_MAXIMIZE
+            || (wp.showCmd == SW_SHOWMINIMIZED
+                && (wp.flags & WPF_RESTORETOMAXIMIZED)));
+    }
+
+    if (RegKey.Create(HKEY_CURRENT_USER, L"Software\\ReactOS\\rapps", NULL,
+        REG_OPTION_NON_VOLATILE, KEY_WRITE, NULL, NULL) != ERROR_SUCCESS)
+    {
+        return FALSE;
+    }
+
+    BOOL bRet = SaveAllSettings(RegKey, *pSettingsInfo);
```

Comment:
you can just `return SaveAllSettiongs(....` here, because the CRegKey will handle closing when going out of scope